packages/flutter/lib/src/widgets/undo_history.dart

```diff
@@ -0,0 +1,447 @@
+      widget.controller.onUndo.addListener(undo);
+      widget.controller.onRedo.addListener(redo);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.value.removeListener(_push);
+    widget.focusNode.removeListener(_handleFocus);
+    widget.controller.onUndo.removeListener(undo);
+    widget.controller.onRedo.removeListener(redo);
+    _throttleTimer?.cancel();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Actions(
+      actions: <Type, Action<Intent>>{
+        UndoTextIntent: Action<UndoTextIntent>.overridable(context: context, defaultAction: CallbackAction<UndoTextIntent>(onInvoke: _undoFromIntent)),
+        RedoTextIntent: Action<RedoTextIntent>.overridable(context: context, defaultAction: CallbackAction<RedoTextIntent>(onInvoke: _redoFromIntent)),
+      },
+      child: widget.child,
+    );
+  }
+}
+
+/// Represents whether the current undo stack can undo or redo.
+@immutable
+class UndoHistoryValue {
+  /// Creates a value for whether the current undo stack can undo or redo.
+  ///
+  /// The [canUndo] and [canRedo] arguments must have a value, but default to
+  /// false.
+  const UndoHistoryValue({this.canUndo = false, this.canRedo = false});
+
+  /// A value corresponding to an undo stack that can neither undo nor redo.
+  static const UndoHistoryValue empty = UndoHistoryValue();
+
+  /// Whether the current undo stack can perform an undo operation.
+  final bool canUndo;
+
+  /// Whether the current undo stack can perform a redo operation.
+  final bool canRedo;
+
+  @override
+  String toString() => '${objectRuntimeType(this, 'UndoHistoryValue')}(canUndo: $canUndo, canRedo: $canRedo)';
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other))
+      return true;
+    return other is UndoHistoryValue && other.canUndo == canUndo && other.canRedo == canRedo;
+  }
+
+  @override
+  int get hashCode => Object.hash(
+        canUndo.hashCode,
+        canRedo.hashCode,
+      );
+}
+
+/// A controller for the undo history, for example for an editable text field.
+///
+/// Whenever a change happens to the underlying value that the [UndoHistory]
+/// widget tracks, that widget updates the [value] and the controller notifies
+/// it's listeners. Listeners can then read the canUndo and canRedo
+/// properties of the value to discover whether [undo] or [redo] are possible.
+///
+/// The controller also has [undo] and [redo] methods to modify the undo
+/// history.
+///
+/// {@tool dartpad}
+/// This example creates a [TextField] with an [UndoHistoryController]
+/// which provides undo and redo buttons.
+///
+/// ** See code in examples/api/lib/widgets/undo_history/undo_history_controller.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+/// * [EditableText], which uses the [UndoHistory] widget and allows
+///   control of the underlying history using an [UndoHistoryController].
+class UndoHistoryController extends ValueNotifier<UndoHistoryValue> {
```

Comment:
You are right that the `UndoHistoryValue` is more like a state than a value, so it is different from the `TextEditingController` in that way. That value is important for propagating the current undo/redo state back to the UI so that custom undo/redo interfaces can be built. The other two ChangeNotifiers are for sending events to allow programmatic undo/redo.

I don't love the idea of adding it to TextEditingController, or at least not _only_ there, so that this can remain more generic: for example, someone could build a drawing interface with it and undo/redo lines on the drawing.

I could get on board with Intents, since we already use those for the keyboard shortcuts, though I agree that might be more confusing.

Reply:
Reading through this again I think I'm onboard with UndoHistoryController as-is.  It makes a lot of sense in the example you included in the examples directory.